src/platform/crypt_openssl.c

```diff
@@ -69,15 +131,83 @@ CxPlatCryptInitialize(
     //
     // LINUX_TODO:Add Check for openssl library QUIC support.
     //
+#endif
+
+    //
+    // Preload cryptographic algorithms
+    //
+#ifdef IS_OPENSSL_3
+    EVP_MAC *mac = NULL;
+
+    if (!CxPlatLoadCipher("AES-128-GCM", &CXPLAT_AES_128_GCM_ALG_HANDLE))
+        goto Error;
+    if (!CxPlatLoadCipher("AES-256-GCM", &CXPLAT_AES_256_GCM_ALG_HANDLE))
+        goto Error;
+    if (!CxPlatLoadCipher("AES-128-ECB", &CXPLAT_AES_128_ECB_ALG_HANDLE))
+        goto Error;
+    if (!CxPlatLoadCipher("AES-256-ECB", &CXPLAT_AES_256_ECB_ALG_HANDLE))
+        goto Error;
+    if (!CxPlatLoadCipher("ChaCha20-Poly1305", &CXPLAT_CHACHA20_POLY1305_ALG_HANDLE))
+        goto Error;
+
+    mac = EVP_MAC_fetch(NULL, "HMAC", "");
+    if (mac == NULL) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            ERR_get_error(),
+            "EVP_MAC_fetch failed");
+        goto Error;
+    }
+    if (!CxPlatLoadHMACCTX(mac, "sha256", &CXPLAT_HMAC_SHA256_CTX_HANDLE))
+        goto Error;
+    if (!CxPlatLoadHMACCTX(mac, "sha384", &CXPLAT_HMAC_SHA384_CTX_HANDLE))
+        goto Error;
+    if (!CxPlatLoadHMACCTX(mac, "sha512", &CXPLAT_HMAC_SHA512_CTX_HANDLE))
+        goto Error;
+    EVP_MAC_free(mac);
 
     return QUIC_STATUS_SUCCESS;
+
+Error:
+    EVP_MAC_free(mac);
+    CxPlatCryptUninitialize();
+    return QUIC_STATUS_OUT_OF_MEMORY;
+#else // ! IS_OPENSSL_3
+    CXPLAT_AES_128_GCM_ALG_HANDLE = (EVP_CIPHER *)EVP_aes_128_gcm();
+    CXPLAT_AES_256_GCM_ALG_HANDLE = (EVP_CIPHER *)EVP_aes_256_gcm();
+    CXPLAT_AES_128_ECB_ALG_HANDLE = (EVP_CIPHER *)EVP_aes_128_ecb();
+    CXPLAT_AES_256_ECB_ALG_HANDLE = (EVP_CIPHER *)EVP_aes_256_ecb();
+    CXPLAT_CHACHA20_POLY1305_ALG_HANDLE = (EVP_CIPHER *)EVP_chacha20_poly1305();
```

Comment:
Seems the chacha20 unit tests are failing (cross-platform). Here's one example:
```
##[error][03/28/2022 11:26:42] CryptTest.WellKnownChaChaPolyv1 failed:
Note: Google Test filter = CryptTest.WellKnownChaChaPolyv1
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from CryptTest
[ RUN      ] CryptTest.WellKnownChaChaPolyv1
/home/vsts/work/1/s/src/platform/unittest/CryptTest.cpp:327: Failure
Expected equality of these values:
  0
  memcmp(HpMask, ExpectedHpMask.Data, ExpectedHpMask.Length)
    Which is: -1
[  FAILED  ] CryptTest.WellKnownChaChaPolyv1 (0 ms)
```

Reply:
This is weird, it's not clear why the changes I made in either use of this cipher would impact this test.  Not using a fetched cipher results in the same error - this diff:
```diff
diff --git a/src/platform/crypt_openssl.c b/src/platform/crypt_openssl.c
index d7d25e6c..c53d826d 100644
--- a/src/platform/crypt_openssl.c
+++ b/src/platform/crypt_openssl.c
@@ -241,7 +241,7 @@ CxPlatKeyCreate(
         Aead = CXPLAT_AES_256_GCM_ALG_HANDLE;
         break;
     case CXPLAT_AEAD_CHACHA20_POLY1305:
-        Aead = CXPLAT_CHACHA20_POLY1305_ALG_HANDLE;
+        Aead = (EVP_CIPHER *)EVP_chacha20_poly1305(); /*CXPLAT_CHACHA20_POLY1305_ALG_HANDLE;*/
         break;
     default:
         Status = QUIC_STATUS_NOT_SUPPORTED;
@@ -515,7 +515,7 @@ CxPlatHpKeyCreate(
         Aead = CXPLAT_AES_256_ECB_ALG_HANDLE;
         break;
     case CXPLAT_AEAD_CHACHA20_POLY1305:
-        Aead = CXPLAT_CHACHA20_POLY1305_ALG_HANDLE;
+        Aead = (EVP_CIPHER *)EVP_chacha20_poly1305(); /*CXPLAT_CHACHA20_POLY1305_ALG_HANDLE;*/;
         break;
     default:
         Status = QUIC_STATUS_NOT_SUPPORTED;
```

I switch to using `EVP_CIPHER_CTX_set_params()` instead of `EVP_CIPHER_CTX_ctrl()` but that doesn't make a difference either, but it demonstrates that it isn't the shim code between the latter and the former that's broken.

Reply:
Could it be your removal of `OPENSSL_init_ssl(OPENSSL_INIT_LOAD_CONFIG, NULL)`? The tests are still failing with your latest changes.

Reply:
@paulidale not sure if you saw my previous comment, but I am guessing your removal of `OPENSSL_init_ssl` is what caused the failure for chacha20 tests.

Reply:
No, I missed it.  The configuration should load automatically, but it's not problem to put the init call back.

Reply:
Well it seems that the tests are still failing even with the latest changes. Really weird!

Reply:
Not too surprising: the init call and config file load are automatic in 3.0 (& 1.1.1).

What is surprising is that a crypto algorithm is failing.  The reason eludes me currently.